### PR TITLE
feat: Create Bidirectional Dijkstra Algorithm

### PR DESCRIPTION
**Pull request by Marinovsky**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1410"></a>

### CODE REVIEW

File: graph/bidirectional_dijkstra.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+            return Shortest_Path_Distance(workset, dist);
+        }
+        visited[currentNode] = 1;
+    }
+    return -1;
+}
+}  // namespace graph
+
+/** Function to test the Algorithm */
+void tests() {
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:186]:
```suggestion
static void tests() {
```

File: graph/bidirectional_dijkstra.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        if (visited[currentNode] == 1) {
+            return Shortest_Path_Distance(workset, dist);
+        }
+        visited[currentNode] = 1;
+    }
+    return -1;
+}
+}  // namespace graph
+
+/** Function to test the Algorithm */
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:185]:
```suggestion
/**
 * @brief Function to test the Algorithm
 * @returns 0 on exit
 */
```

File: graph/bidirectional_dijkstra.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    graph::addEdge(&adj2_1, &adj2_2, 3, 4, 4);
+
+    s = 1, t = 5;
+    std::cout << "Initiating Test 3..." << std::endl;
+    assert(graph::Bidijkstra(&adj2_1, &adj2_2, s - 1, t - 1) == 6);
+    std::cout << "Test 3 Passed..." << std::endl;
+    std::cout << "All Test Passed..." << std::endl << std::endl;
+}
+
+/** Main function */
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:228]:
```suggestion
/** 
 * @brief Main function
 * @returns 0 on exit
 */
```

File: graph/bidirectional_dijkstra.cpp [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+#include <utility>
+#include <vector>
+
+constexpr int64_t INF = std::numeric_limits<int64_t>::max();
+
+/**
+ * @namespace graph
+ * @brief Graph Algorithms
+ */
+
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:33]:
```suggestion
```

File: graph/bidirectional_dijkstra.cpp [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+ *
+ * @details
+ * This is basically the same Dijkstra Algorithm but faster because it goes from
+ * the source to the target and from target to the source and stops when
+ * finding a vertex visited already by the direct search or the reverse one.
+ * Here some simulations of it:
+ * https://www.youtube.com/watch?v=DINCL5cd_w0&t=24s
+ */
+
+#include <math.h>
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:16]:
```suggestion
#include <cmath>
```

File: graph/bidirectional_dijkstra.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+#include <vector>
+
+constexpr int64_t INF = std::numeric_limits<int64_t>::max();
+
+/**
+ * @namespace graph
+ * @brief Graph Algorithms
+ */
+
+namespace graph {
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:34]:
```suggestion
namespace graph {
/**
 * @namespace bidirectional_dijkstra
 * @brief Functions for [Bidirectional Dijkstra Shortest Path]
 * (https://www.coursera.org/learn/algorithms-on-graphs/lecture/7ml18/bidirectional-dijkstra) algorithm
 */
namespace bidirectional_dijkstra {
```

File: graph/bidirectional_dijkstra.cpp [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+        /// check if currentNode has already been visited
+        if (visited[currentNode] == 1) {
+            return Shortest_Path_Distance(workset, dist);
+        }
+        visited[currentNode] = 1;
+    }
+    return -1;
+}
+}  // namespace graph
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:183]:
```suggestion
}  // namespace bidirectional_dijkstra
}  // namespace graph
```

File: graph/bidirectional_dijkstra.cpp [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    std::cout << "Initiatinig Predefined Tests..." << std::endl;
+    std::cout << "Initiating Test 1..." << std::endl;
+    std::vector<std::vector<std::pair<int, int>>> adj1_1(
+        4, std::vector<std::pair<int, int>>());
+    std::vector<std::vector<std::pair<int, int>>> adj1_2(
+        4, std::vector<std::pair<int, int>>());
+    graph::addEdge(&adj1_1, &adj1_2, 1, 2, 1);
+    graph::addEdge(&adj1_1, &adj1_2, 4, 1, 2);
+    graph::addEdge(&adj1_1, &adj1_2, 2, 3, 2);
+    graph::addEdge(&adj1_1, &adj1_2, 1, 3, 5);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:196]:
```suggestion
    graph::bidirectional_dijkstra::addEdge(&adj1_1, &adj1_2, 1, 2, 1);
    graph::bidirectional_dijkstra::addEdge(&adj1_1, &adj1_2, 4, 1, 2);
    graph::bidirectional_dijkstra::addEdge(&adj1_1, &adj1_2, 2, 3, 2);
    graph::bidirectional_dijkstra::addEdge(&adj1_1, &adj1_2, 1, 3, 5);
```

File: graph/bidirectional_dijkstra.cpp [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+    int s = 1, t = 3;
+    assert(graph::Bidijkstra(&adj1_1, &adj1_2, s - 1, t - 1) == 3);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:199]:
```suggestion
    assert(graph::bidirectional_dijkstra::Bidijkstra(&adj1_1, &adj1_2, s - 1, t - 1) == 3);
```

File: graph/bidirectional_dijkstra.cpp [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    std::cout << "Test 1 Passed..." << std::endl;
+
+    s = 4, t = 3;
+    std::cout << "Initiating Test 2..." << std::endl;
+    assert(graph::Bidijkstra(&adj1_1, &adj1_2, s - 1, t - 1) == 5);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:204]:
```suggestion
    assert(graph::bidirectional_dijkstra::Bidijkstra(&adj1_1, &adj1_2, s - 1, t - 1) == 5);
```

File: graph/bidirectional_dijkstra.cpp [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        5, std::vector<std::pair<int, int>>());
+    graph::addEdge(&adj2_1, &adj2_2, 1, 2, 4);
+    graph::addEdge(&adj2_1, &adj2_2, 1, 3, 2);
+    graph::addEdge(&adj2_1, &adj2_2, 2, 3, 2);
+    graph::addEdge(&adj2_1, &adj2_2, 3, 2, 1);
+    graph::addEdge(&adj2_1, &adj2_2, 2, 4, 2);
+    graph::addEdge(&adj2_1, &adj2_2, 3, 5, 4);
+    graph::addEdge(&adj2_1, &adj2_2, 5, 4, 1);
+    graph::addEdge(&adj2_1, &adj2_2, 2, 5, 3);
+    graph::addEdge(&adj2_1, &adj2_2, 3, 4, 4);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:219]:
```suggestion
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 1, 2, 4);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 1, 3, 2);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 2, 3, 2);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 3, 2, 1);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 2, 4, 2);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 3, 5, 4);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 5, 4, 1);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 2, 5, 3);
    graph::bidirectional_dijkstra::addEdge(&adj2_1, &adj2_2, 3, 4, 4);
```

File: graph/bidirectional_dijkstra.cpp [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+    s = 1, t = 5;
+    std::cout << "Initiating Test 3..." << std::endl;
+    assert(graph::Bidijkstra(&adj2_1, &adj2_2, s - 1, t - 1) == 6);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:223]:
```suggestion
    assert(graph::bidirectional_dijkstra::Bidijkstra(&adj2_1, &adj2_2, s - 1, t - 1) == 6);
```

File: graph/bidirectional_dijkstra.cpp [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+
+    std::vector<std::vector<std::pair<int, int>>> adj1(
+        vertices, std::vector<std::pair<int, int>>());
+    std::vector<std::vector<std::pair<int, int>>> adj2(
+        vertices, std::vector<std::pair<int, int>>());
+
+    int u = int(), v = int(), w = int();
+    while (edges--) {
+        std::cin >> u >> v >> w;
+        graph::addEdge(&adj1, &adj2, u, v, w);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:246]:
```suggestion
        graph::bidirectional_dijkstra::addEdge(&adj1, &adj2, u, v, w);
```

File: graph/bidirectional_dijkstra.cpp [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    int s = int(), t = int();
+    std::cin >> s >> t;
+    int dist = graph::Bidijkstra(&adj1, &adj2, s - 1, t - 1);
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:251]:
```suggestion
    int dist = graph::bidirectional_dijkstra::Bidijkstra(&adj1, &adj2, s - 1, t - 1);
```

File: graph/bidirectional_dijkstra.cpp [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+#include <math.h>
+#include <cassert>
+#include <cstdio>
+#include <iostream>
+#include <limits>
+#include <set>
+#include <unordered_map>
+#include <queue>
+#include <utility>
+#include <vector>
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:25]: Add a brief description of what the library/header adds (see the example below).

        #include <assert>     /// for assert
        #include <iostream>  /// for io operations

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good code.

**Comment by Marinovsky** [comment]
Hey, I was trying to commit the changes you mentioned but I don't know why some checks were not successful, and I this is my first pull request, so can I close this pull request apply the changes mentioned, and make a new pull request?

**Comment by Panquesito7** [comment]
No, it's OK to implement the same changes in the same PR for easier reviewing. 🙂
I'll check the continuous integration errors. Seems to be something I messed up. 😄

File: graph/bidirectional_dijkstra.cpp [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+ * 
+ * @param workset_ vertices visited in the search
+ * @param distance_ vector of distances from the source to the target and
+ * from the target to the source
+ * 
+ */
+int Shortest_Path_Distance(std::vector<int> workset_,
+                  std::vector<std::vector<int64_t>> distance_) {
+    int64_t distance = INF;
+    for (int i = 0; i < workset_.size(); i++) {
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:67]:
```suggestion
    for (int i : workset_) {
```

File: graph/bidirectional_dijkstra.cpp [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+/**
+ *@brief This function returns the shortest distance from the source 
+ * to the target if there is path between vertices 's' and 't'.
+ * 
+ * @param workset_ vertices visited in the search
+ * @param distance_ vector of distances from the source to the target and
+ * from the target to the source
+ * 
+ */
+int Shortest_Path_Distance(std::vector<int> workset_,
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:64]:
```suggestion
int Shortest_Path_Distance(const std::vector<int> &workset_,
```

File: graph/bidirectional_dijkstra.cpp [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+            }
+        }
+        // store the processed node/ vertex
+        workset.push_back(currentNode);
+
+        /// check if currentNode has already been visited
+        if(visited[currentNode] == 1){
+            return Shortest_Path_Distance(workset, dist);
+        }
+        visited[currentNode] = 1;
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:153]:
```suggestion
        visited[currentNode] = true;
```

File: graph/bidirectional_dijkstra.cpp [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+            }
+        }
+        // store the processed node/ vertex
+        workset.push_back(currentNode);
+
+        /// check if currentNode has already been visited
+        if (visited[currentNode] == 1) {
+            return Shortest_Path_Distance(workset, dist);
+        }
+        visited[currentNode] = 1;
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:184]:
```suggestion
        visited[currentNode] = true;
```

File: graph/bidirectional_dijkstra.cpp [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    return -1;
+}
+}  // namespace bidirectional_dijkstra
+}  // namespace graph
+
+/** Function to test the Algorithm */
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:191]:
```suggestion
/**
 * @brief Function to test the
 * provided algorithm above
 * @returns void
 */
```

File: graph/bidirectional_dijkstra.cpp [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+#include <cmath>
+#include <cassert>
+#include <cstdio>
+#include <iostream>
+#include <limits>
+#include <set>
+#include <unordered_map>
+#include <queue>
+#include <utility>
+#include <vector>
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:25]: Add a brief description of what the header/library adds (see the example below).
```suggestion
#include <cmath>          /// for
#include <cassert>        /// for assert
#include <cstdio>         /// for
#include <iostream>       /// for io operations
#include <limits>         /// for
#include <set>            /// for
#include <unordered_map>  /// for
#include <queue>          /// for
#include <utility>        /// for
#include <vector>         /// for std::vector
```

**Comment by Marinovsky** [comment]
I think it is already done

File: graph/bidirectional_dijkstra.cpp [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+namespace bidirectional_dijkstra {
+/**
+ * @brief Function that add edge between two nodes or vertices of graph
+ *
+ * @param adj1 adjacency list for the direct search
+ * @param adj2 adjacency list for the reverse search
+ * @param u any node or vertex of graph
+ * @param v any node or vertex of graph
+ */
+void addEdge(std::vector<std::vector<std::pair<int, int>>> *adj1,
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:44]: Consider using `uint64_t` or `int64_t`, or their appropriate size ([reference](https://en.cppreference.com/w/cpp/types/integer)). Check other parts of the code.

**Review by Panquesito7 — COMMENTED** [review]
Did a quick test on Gitpod, and there seems to be an error with your program.
After printing the result, it keeps there forever, unless you type something and enter.

Gitpod quick-test

After typing the vertices and edges:

Gitpod quick-test


Could you please fix the algorithm to run properly?

**Comment by Marinovsky** [comment]
> Did a quick test on Gitpod, and there seems to be an error with your program.
> After printing the result, it keeps there forever, unless you type something and enter.
> 
> Gitpod quick-test
> 
> After typing something the vertices and edges:
> 
> Gitpod quick-test
> 
> Could you please fix the algorithm to run properly?

Yeah, I think it is not a problem, after executing the test cases the user can enter a test case so they first enter the number of vertices and edges and then the graph by the number of edges lines in which there are three numbers v, u, and w, representing an edge between the node v and u with weight w. I will describe that in the commentaries.

File: graph/bidirectional_dijkstra.cpp [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+ */
+void addEdge(std::vector<std::vector<std::pair<uint64_t, uint64_t>>> *adj1,
+             std::vector<std::vector<std::pair<uint64_t, uint64_t>>> *adj2,
+             uint64_t u, uint64_t v, uint64_t w) {
+    (*adj1)[u - 1].push_back(std::make_pair(v - 1, w));
+    (*adj2)[v - 1].push_back(std::make_pair(u - 1, w));
+    // (*adj)[v - 1].push_back(std::make_pair(u - 1, w));
+}
+/**
+ *@brief This function returns the shortest distance from the source
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:53]:
```suggestion
 * @brief This function returns the shortest distance from the source
```

File: graph/bidirectional_dijkstra.cpp [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+    std::cout << "All Test Passed..." << std::endl << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    // running predefined tests
+    tests();
```
Comment (Panquesito7) [inline graph/bidirectional_dijkstra.cpp:250]:
```suggestion
    tests();  // running predefined tests
```

**Review by Panquesito7 — APPROVED** [review]
Neat work! Code looks refined and well-documented! 😄 👍

**Comment by Marinovsky** [comment]
Excuse me, Shall I make another modification to the code?

**Review by ayaankhan98 — APPROVED** [review]
Thanks for your contribution @Marinovsky, LGTM!

**Pull request merged**